New Post - GUI Challenges - Building a dialog component (#7659)
GoogleChrome/web.dev: src/site/content/en/blog/building-a-dialog-component/index.md

---
layout: post
title: Building a dialog component
subhead: A foundational overview of how to build color-adaptive, responsive, and accessible mini and mega modals with the `<dialog>` element.
authors:
  - adamargyle

> description: A foundational overview of how to build color-adaptive, responsive, and accessible mini and mega modals with the `<dialog>` element.

Author: ```suggestion
description: A foundational overview of how to build color-adaptive, responsive, and accessible mini and mega modals with the <dialog> element.
```

colors](https://codepen.io/argyleink/pen/KKvRORE) that adapt automatically to
light and dark system preferences, similar to using the `color-scheme`. These
are great for creating layers in a design and I love using color to help
visually support this appearance of layer surfaces. The background color is
`var(--surface-1)`; to sit on top of that layer, use `var(--surface-2)`:

```css
dialog {
  …
  background: var(--surface-2);
  color: var(--text-1);
}

@media (prefers-color-scheme: dark) {
  dialog {
    border-block-start: var(--border-size-1) solid var(--surface-3);
  }
}
```

More adaptive colors will be added later for child elements, such as the header
and footer. I consider them extra for a dialog element, but really important in
making a compelling and well designed dialog design.

{% Aside %}
For a deep dive on adaptive color systems, 
see the [building a color scheme](/building-a-color-scheme/) GUI Challenge.
{% endAside %}

#### Responsive dialog sizing

The dialog defaults to delegating its size to its contents, which is generally
great. My goal here is to constrain the
[`max-inline-size`](https://developer.mozilla.org/docs/Web/CSS/max-inline-size)

> to a readable size (--size-content-3 = 60ch) or 90% of the viewport width. This

Reviewer: ```suggestion
to a readable size (`--size-content-3` = `60ch`) or 90% of the viewport width. This
```

The following styles fix the dialog element to the window, stretching it to each
corner, and uses `margin: auto` to center the content:

```css
dialog {
  …
  margin: auto;
  padding: 0;
  position: fixed;
  inset: 0;
  z-index: var(--layer-important);
}
```

<figure>
{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/fsjCyaR7hwOKuSuL0jPk.mp4",
  autoplay="true",
  loop="true",
  muted="true" 
%}
</figure>

##### Mobile mega dialog styles

On small viewports, I style this full page mega modal a little differently. I
set the bottom margin to `0`, which brings the dialog content to the bottom of
the viewport. With a couple of style adjustments, I can turn the dialog into an
actionsheet, closer to the user's thumbs:

```css
@media (max-width: 768px) {
  dialog[modal-mode="mega"] {
    margin-block-end: 0;
    border-end-end-radius: 0;
    border-end-start-radius: 0;
  }
}
```

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/pmvxlxa2b3SnTnCow5gi.png", 
  alt="Screenshot of devtools overlaying margin spacing 
  on both the desktop and mobile mega dialog while open.", 
  width="800", 
  height="484" 
%}

##### Mini dialog positioning

> When using a desktops larger viewport, I chose to position the mini dialogs over

Reviewer: ```suggestion
When using a larger viewport such as on a desktop computer, I chose to position the mini dialogs over
```

#### Make it pop

Last, add some flair to the dialog so it looks like a soft surface sitting far
above the page. The softness is achieved by rounding the corners of the dialog.
The depth is achieved with one of Open Props’ carefully crafted [shadow
props](https://open-props.style/#shadows):

```css
dialog {
  …
  border-radius: var(--radius-3);
  box-shadow: var(--shadow-6);
}
```

#### Customizing the backdrop pseudo element

I chose to work very lightly with the backdrop, only adding a blur effect with
[`backdrop-filter`](https://developer.mozilla.org/docs/Web/CSS/backdrop-filter)
to the mega dialog:

{% BrowserCompat 'css.properties.backdrop-filter' %}

```css
dialog[modal-mode="mega"]::backdrop {
  backdrop-filter: blur(25px);
}
```

I also chose to put a transition on `backdrop-filter`, in the hope that browsers
will allow transitioning the backdrop element in the future:

```css
dialog::backdrop {
  transition: backdrop-filter .5s ease;
}
```

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/lufi4kzHHbEnnEVbZoFD.png", 
  alt="Screenshot of the mega dialog overlaying a blurred background of colorful avatars.", 
  width="800", 
  height="509" 
%}

### Styling extras

> I call this section “extras” because it has more to do with my dialog element

Reviewer: ```suggestion
I call this section "extras" because it has more to do with my dialog element
```

description: A foundational overview of how to build color-adaptive, responsive, and accessible mini and mega modals with the <dialog> element.
date: 2022-04-13
thumbnail: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/7l53GAQu1kmIuqOmTjRI.png
hero: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/iVRNS1o7v27mmBPQiDdk.png
alt: Preview of the mini and mega dialogs in both light and dark themes.
tags:
  - blog
  - css
  - html
  - javascript
---

In this post I want to share my thoughts on how to build color-adaptive,
responsive, and accessible mini and mega modals with the `<dialog>` element.
[Try the demo](https://gui-challenges.web.app/dialog/dist/) and [view the
source](https://github.com/argyleink/gui-challenges)!

<figure data-size="full">
  <style style="display: none">
    .adjusted-aspect-ratio {
      aspect-ratio: 4/3;
    }
  </style>
  {% Video
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/rx8k6w5iQnEOwqOoUNNG.mp4",
    autoplay="true",
    loop="true",
    muted="true",
    class="adjusted-aspect-ratio"
  %}
  <figcaption>
    Demonstration of the mega and mini dialogs in their light and dark themes.
  </figcaption>
</figure>

If you prefer video, here's a YouTube version of this post:

> {% YouTube 'TD1QxlGgIPk' %}

Reviewer: ```suggestion
{% YouTube 'GDzzIlRhEzM' %}
```
